tests/test_writer.py

```diff
@@ -2480,3 +2480,14 @@ def test_append_pdf_with_dest_without_page(caplog):
     writer.append(reader)
     assert "/__WKANCHOR_8" not in writer.named_destinations
     assert len(writer.named_destinations) == 3
+
+
+def test_stream_not_closed():
+    """Tests for #2905"""
+    src = RESOURCE_ROOT / "pdflatex-outline.pdf"
+    with NamedTemporaryFile() as tmp:
+        with PdfReader(src) as reader, PdfWriter() as writer:
+            for i in range(4):
+                writer.add_page(reader.pages[i])
+            writer.write(tmp)
+        assert not tmp.file.closed
```

Comment:
it might be great to also have a test for where the automatic write at the closure of the context will be done

Reply:
I think I managed to add a test for this.  It was a bit confusing because of the double-construct that happens in `__enter__()`.  I spent a fair amount of time trying to understand the `clone_from` logic before I realized that everything from the first `__init__()` is thrown away except for `temp_fileobj`.